File: 0535-encode-and-decode-tinyurl/0535-encode-and-decode-tinyurl.cpp
class Solution {
public:
    unordered_map<string, string> map;
    int count = 0;

    // Encodes a URL to a shortened URL.
    string encode(string longUrl) {
        string key = "http://tinyurl.com/" + count++;
        map[key] = longUrl;
        return key;
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl) {
        return map[shortUrl];
    }
};

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));

File: 0535-encode-and-decode-tinyurl/0535-encode-and-decode-tinyurl.cs
public class Codec {
    private readonly Dictionary<string, string> hashmap;

    public Codec() {
        hashmap = new Dictionary<string, string>();
    }

    // Encodes a URL to a shortened URL
    public string encode(string longUrl) {
        string key;
        do {
            key = Convert.ToBase64String(Guid.NewGuid().ToByteArray())
                .Replace("+", "-")
                .Replace("/", "_")
                .TrimEnd('=');
        } while(hashmap.ContainsKey(key));

        hashmap[key] = longUrl;
        return key;
    }

    // Decodes a shortened URL to its original URL.
    public string decode(string shortUrl) {
        return hashmap[shortUrl];
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));